actors/builtin/reward/reward_state.go

```diff
@@ -3,305 +3,68 @@ package reward
+	// CumsumRealized is cumulative sum of network power capped by BalinePower(epoch)
+	CumsumRealized Spacetime
 
-	SimpleSupply   abi.TokenAmount // current supply
-	BaselineSupply abi.TokenAmount // current supply
+	// EffectiveNetworkTime is ceiling of real effective network time `theta` based on
+	// CumsumBaselinePower(theta) == CumsumRealizedPower
+	// Theta captures the notion of how much the network has progressed in its baseline
+	// and in advancing network time.
+	EffectiveNetworkTime abi.ChainEpoch
 
-	// The reward to be paid in total to block producers, if exactly the expected number of them produce a block.
+	// The reward to be paid in per WinCount to block producers.
 	// The actual reward total paid out depends on the number of winners in any round.
-	// This is computed at the end of the previous epoch, and should really be called ThisEpochReward.
 	ThisEpochReward abi.TokenAmount
 
-	// The count of epochs for which a reward has been paid.
-	// This should equal the number of non-empty tipsets after the genesis, aka "chain height".
-	RewardEpochsPaid abi.ChainEpoch
+	// Epoch tracks for which epoch the Reward was computed
+	Epoch abi.ChainEpoch
 }
 
-type AddrKey = adt.AddrKey
-
-// These numbers are placeholders, but should be in units of attoFIL, 10^-18 FIL
-var SimpleTotal = big.Mul(big.NewInt(100e6), big.NewInt(1e18))   // 100M for testnet, PARAM_FINISH
-var BaselineTotal = big.Mul(big.NewInt(900e6), big.NewInt(1e18)) // 900M for testnet, PARAM_FINISH
-
-func ConstructState(currRealizedPower *abi.StoragePower) *State {
+func ConstructState(currRealizedPower abi.StoragePower) *State {
 	st := &State{
-		BaselinePower:        big.Zero(),
-		RealizedPower:        big.Zero(),
 		CumsumBaseline:       big.Zero(),
 		CumsumRealized:       big.Zero(),
-		EffectiveNetworkTime: big.Zero(),
+		EffectiveNetworkTime: 0,
 
-		SimpleSupply:     big.Zero(),
-		BaselineSupply:   big.Zero(),
-		ThisEpochReward:  big.Zero(),
-		RewardEpochsPaid: 0,
+		ThisEpochReward: big.Zero(),
+		Epoch:           -1,
 	}
-	st.updateToNextEpochReward(currRealizedPower)
+
+	st.updateToNextEpochWithReward(currRealizedPower)
 
 	return st
 }
 
-// Takes in a current realized power for a reward epoch and computes
-// and updates reward state to track reward for the next epoch
-func (st *State) updateToNextEpochReward(currRealizedPower *abi.StoragePower) {
-	st.RealizedPower = *currRealizedPower
-
-	st.BaselinePower = st.newBaselinePower()
-	st.CumsumBaseline = big.Add(st.CumsumBaseline, st.BaselinePower)
+// Takes in current realized power and updates internal state
+// Used for update of internal state during null rounds
+func (st *State) updateToNextEpoch(currRealizedPower abi.StoragePower) {
+	st.Epoch++
 
-	// Cap realized power in computing CumsumRealized so that progress is only relative to the current epoch.
-	cappedRealizedPower := big.Min(st.BaselinePower, st.RealizedPower)
+	cappedRealizedPower := big.Min(BaselinePowerAt(st.Epoch), currRealizedPower)
```

Comment:
@zixuanzh I don't want to hold up the PR for this (which existed already), but this doesn't match my expectation of how this should work, so want to check whether it matches yours. As I understand this, if at any point the realized spacetime exceeds the baseline, we are forgetting about the excess. This means that if later the realized drops below the baseline, rewards are immediately reduced even if, in reality, the integral of realised power over time exceeds the integral of baseline.

This is as opposed to the approach of calculating the realized cumsum accurately, and then taking the `min` when using it to calculate reward.

Reply:
Yes, we are explicitly capping the realized spacetime at the instantaneous baseline and it is also correct that if the realized baseline immediately drops below the baseline and the baseline share of the reward will reduce (this is part of why we have a proportion allocation to simple minting). This will happen when storage is pulled out from the network. The Min is applied at the instantaneous baseline, not at the limit. Otherwise, we will see very unstable block rewards if the min is at the cumsum. cc @davidad